Add bbg for ballerina GKE deployment (#1)
ballerina-guides/ballerina-gke-deployment: README.md

[![Build Status](https://travis-ci.org/ballerina-guides/ballerina-gke-deployment.svg?branch=master)](https://travis-ci.org/ballerina-guides/ballerina-gke-deployment)

# Ballerina Deployment with Google Cloud Platoform Kubernetes Engine

[Google Kubernetes Engine (GKE)](https://cloud.google.com/kubernetes-engine/) is a powerful cluster manager and orchestration system for running your Docker containers

> In this guide you will learn about building a Ballerina service and deploying it on Google Kubernetes Enginer (GKE).

> The following are the sections available in this guide.

Reviewer: Are we not covering observability here?

Author: No, we are focusing on a simple guide on a remote deployment on Google Kubernetes Engine. Also I don't think it makes sense set up the observability stuff similarly to how we did locally. 
May be we can do it properly as an improvement later.

- [Prerequisites](#prerequisites)
- [Implementation](#implementation)
- [Deployment](#deployment)
- [Testing](#testing)

## What you’ll build

In this guide, you will build a simple Ballerina Hello World service, and you will deploy this service on GKE.

## Compatibility

| Ballerina Language Version
| --------------------------
| 0.990.0

## Prerequisites

- [Ballerina Distribution](https://ballerina.io/learn/getting-started/)
- A Text Editor or an IDE
- [Docker](https://docs.docker.com/engine/installation/)
- [kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/)
- [Docker Hub Account](https://hub.docker.com/)
- [Google Cloud Platform account](https://cloud.google.com/)

### Optional requirements

- Ballerina IDE plugins ([IntelliJ IDEA](https://plugins.jetbrains.com/plugin/9520-ballerina), [VSCode](https://marketplace.visualstudio.com/items?itemName=WSO2.Ballerina))

## Implementation

### Create the project structure

Ballerina is a complete programming language that can have any custom project structure that you wish. Although the language allows you to have any module structure, use the following module structure for this project to follow this guide.

> ```

Reviewer: Add a line break between text and the back ticks. There are some rendering issues that occur sometimes when this line break does not exist.

ballerina-gke-deployment
 └── guide
      └── hello_world_service.bal
```

> - Create the above directories in your local machine and also create empty `.bal` file.

Reviewer: ```suggestion
- Create the above directories in your local machine and also create an empty `.bal` file.
```

> - Then open the terminal and navigate to `ballerina-gke-deployment/guide` and run Ballerina project initializing toolkit.

Reviewer: ```suggestion
- Open the terminal and navigate to `ballerina-gke-deployment/guide` and run Ballerina project initializing toolkit.
```

            log:printError("Error sending response", err = result);
        }
    }
}
```

Now you can add the Kubernetes annotations that are required to generate the Kubernetes deployment artifacts.

```ballerina
import ballerina/http;
import ballerina/log;
import ballerinax/kubernetes;

@kubernetes:Service {
    name:"hello-world",
    serviceType:"LoadBalancer",
    port:80
}
listener http:Listener httpListener = new(9090);

@kubernetes:Deployment {
    enableLiveness:true,
    image:"<username>/hello_world_service:latest",
    push:true,
    username:"<username>",
    password:"<password"
}
// By default, Ballerina exposes a service via HTTP/1.1.
service hello on httpListener {

    // Invoke all resources with arguments of server connector and request.
    resource function sayHello(http:Caller caller, http:Request req) {
        http:Response res = new;

        // Use a util method to set a string payload.
        res.setPayload("Hello, World!");

        // Send the response back to the caller.
        var result = caller->respond(res);
        if (result is error) {
            log:printError("Error sending response", err = result);
        }
    }
}

```

> We will be building a Docker image here and publishing it to Docker Hub. This is required, since we cannot simply have the Docker image in the local registry, and run the Kubernetes applicates in GKE, where it needs to have access to the Docker image in a globally accessible location. For this, an image name should be given in the format $username/$image_name in the "image" property, and "username" and "password" properties needs to contain the Docker Hub account username and password respectively. The property "push" is set to "true" to signal the build process to push the build Docker image to Docker Hub.

Reviewer: ```suggestion
We will be building a Docker image here and publishing it to Docker Hub. This is required, since we cannot simply have the Docker image in the local registry, and run the Kubernetes applicates in GKE, where it needs to have access to the Docker image in a globally accessible location. For this, an image name should be given in the format `$username/$image_name` in the `image` property, and `username` and `password` properties needs to contain the Docker Hub account username and password respectively. The property "push" is set to "true" to signal the build process to push the build Docker image to Docker Hub.

```

Your current configuration has been set to: [default]

You can skip diagnostics next time by using the following flag:
  gcloud init --skip-diagnostics

Network diagnostic detects and fixes local network connection issues.
Checking network connection...done.
Reachability Check passed.
Network diagnostic passed (1/1 checks passed).

You must log in to continue. Would you like to log in (Y/n)?

Your browser has been opened to visit:

    https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&prompt=select_account&response_type=code&client_id=32555940559.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fappengine.admin+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcompute+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Faccounts.reauth&access_type=offline

```

Once you click on the provided link, you will be prompted to login to your Google account and authorize.

Once authentication flow completes, you will be prompted to create a new project or choose an existing project as the

> current project. Here, we are creating a project named "BallerinaDemo"

Reviewer: Remove line break and add a period at the end of the sentence. Apply elsewhere if applicable.

hyphens) in length and start with a lowercase letter. BallerinaDemo
Your current project has been set to: [BallerinaDemo].

Not setting default zone/region (this feature makes it easier to use
[gcloud compute] by setting an appropriate default value for the
--zone and --region flag).
See https://cloud.google.com/compute/docs/gcloud-compute section on how to set
default compute region and zone manually. If you would like [gcloud init] to be
able to do this for you the next time you run it, make sure the
Compute Engine API is enabled for your project on the
https://console.developers.google.com/apis page.

Created a default .boto configuration file at [/home/manurip/.boto]. See this file and
[https://cloud.google.com/storage/docs/gsutil/commands/config] for more
information about configuring Google Cloud Storage.
Your Google Cloud SDK is configured and ready to use!
```

With the following command you can list the projects.

```bash
$ gcloud projects list
PROJECT_ID                NAME              PROJECT_NUMBER
ballerinademo-225007      BallerinaDemo     1036334079773
```

- Create the Kubernetes cluster

Next step is creating a kubernetes cluster in the project we just created.

> With a command similar to below, you can create a cluster with minimal resources.

Reviewer: ```suggestion
With a command similar to the following, you can create a cluster with minimal resources.
```
